README.md
# GRAD10S

## Logo:
![](screenshots/gradios.png)

## Descripcion:
Juego basado en Gradius, popular videojuego de arcade. Gradius es un matamarcianos lanzado por Konami en 1985.

## Version 1.1 Lanzada!

agregamos las bombas y las maquinas que no dejan de apuntar al personaje para destruirlo de un bombazo. Esperamos les guste.

![4](screenshots/ss4.jpg)

## Screenshot:

![1](screenshots/ss1.png)
![2](screenshots/ss2.png)
![3](screenshots/ss3.png)

## Musica: 
- Realizada con BoscaCeoil.

## Sonidos:
- Realizados con Bfxr.

## Imagenes y assets:
- '[M13] Cave Tileset', Autor: Beast, https://opengameart.org/content/m13-cave-tileset.
- 'STARLINK BATTLE FOR ATLAS SPACESHIP SPACE BATTLE', sin autor, https://anoncraft.com/starlink-battle-for-atlas-spaceship-space-battle-21216/.
- 'Azul Rojo Galaxy Fantasía Espacio fondo de pantalla', sin autor, http://www.hdfondos.eu/preview/65932/1024/768.
- Sin titulo, Tony Rivera, https://www.clarin.com/deportes/futbol-internacional/diego-maradona-despues-triunfo-dorados-vendo-humo-vendo-trabajo_0_vGCzhB6lP.html.
- La nave, los asteroides y el disparo fueron creados con Adobe Illustrator y las imagenes modificadas con Adobe Photoshop.

## Autor: 
- Mariano Martinez.
## Colaborador:
- PatoDL

## Contacto: 
### Mariano:
- mariano_mcr92@hotmail.com
- marianomcr92@gmail.com

### Pato:
- github.com/PatoDL
- patodl.itch.io
- ig: @pato.dl


CHANGELOG.md
# Changelog

## [1.1] 2018-11-1
### Added
- se agregaron las bombas como forma de ataque que solo afecta a las maquinas
- se agregaron nuevos enemigos "maquinas" que solo pueden ser destruidos con las bombas

## [1.0] 2018-10-29
### Added
- Se agrego la imagen de fondo en el menu.
-Se agregaron las funciones ResetMeteors y ResetSpaceship .
-Se realizo el unload al salir de cada pantalla.
-Se agrego en la pantalla de creditos los creditos de las imagenes.
-Se agrego la variable halfScreen para realizar menos calculos.
-Se agregaron archivos a la carpeta "raw".
### Modified
- Se modifico el boton del sonido para que se note cuando esta activado y cuando no.
- Se modifico la posicion de los asteroides para que no queden mal posicionados en el eje Y .
-Se modifico el nombre de la funcion Draw a DrawGameplay para dar mas claridad.

## [0.3] 2018-10-25
### Added 
- Se agrego animacion a la nave
- Se agregaron los archivos parallax, parallax2 y spritesheet a la carpeta res 
- Se genero un efecto parallax con las imagenes agregadas
- Fix en el .gitignore  

## [0.2] 2018-10-18
### Added 
- Se agregaron los archivos enemies.cpp, enemies.h, meteors.cpp y meteors.h
- Los archivos de enemies estan vacios para luego agregar mas enemigos
- Fix en el movimiento de los meteoros
- Fix en condicion de derrota 

## [0.1] 2018-10-16
### Added
- Se agregaron las pantallas game_over, credits, game, menu, pause
- Se agrego el objeto de la nave y se modifico el movimiento para que se fije en los limites de la pantalla

src/Pantallas/menu.cpp
#include "menu.h"

#include <raylib.h>

#include "pause.h"
#include "app.h"
#include "game.h"

using namespace app;
using namespace pause;
using namespace game;

namespace app
{
	namespace menu
	{
		static char text1[] = "JUGAR";
		static char text2[] = "SONIDO ON";
		static char text3[] = "CREDITOS";
		static char text4[] = "SALIR";
		static char text5[] = "V 1.1";
		static char text6[] = "SONIDO OFF";
		static int sizeText2 = 0;
		static int sizeText3 = 0;
		static int text1PositionX = 0;
		static int text1PositionY = 0;
		static int text2PositionX = 0;
		static int text2PositionY = 0;
		static int text3PositionX = 0;
		static int text3PositionY = 0;
		static int text4PositionX = 0;
		static int text4PositionY = 0;
		static int text5PositionX = 0;
		static int text5PositionY = 0;
		static int text6PositionX = 0;
		static int text6PositionY = 0;
		static Vector2 mousePoint;
		static Rectangle rect1;
		static Rectangle rect2;
		static Rectangle rect3;
		static Rectangle rect4;
		static Color colorRect1;
		static Color colorRect2;
		static Color colorRect3;
		static Color colorRect4;
		extern bool cambiarTexto=false;
		static int halfScreenX;
		static int halfScreenY;

		static const int scaleAux1 = 800;
		static const int scaleAux2 = 1600;
		static const int scaleAux3 = 1280;

		static Image menuImage;
		static Texture2D menuImageTexture;
		static float scaleBackground;
		
		void InitMenu()
		{
			menuImage = LoadImage("res/assets/backMenu.png");
			menuImageTexture = LoadTextureFromImage(menuImage);
			halfScreenX = GetScreenWidth() / 2;
			halfScreenY = GetScreenHeight() / 2;

			sizeText2 = (GetScreenWidth() * 20) / scaleAux1;
			sizeText3 = (GetScreenWidth() * 15) / scaleAux1;
			text1PositionX = halfScreenX - MeasureText(text1, sizeText2) / 2;
			text1PositionY = halfScreenY + GetScreenHeight() * 0.0333333;
			text2PositionX = halfScreenX - MeasureText(text2, sizeText2) / 2;
			text2PositionY = halfScreenY + GetScreenHeight() * 0.1333333;
			text3PositionX = halfScreenX - MeasureText(text3, sizeText2) / 2;
			text3PositionY = halfScreenY + GetScreenHeight() * 0.2333333;
			text4PositionX = halfScreenX - MeasureText(text4, sizeText2) / 2;
			text4PositionY = halfScreenY + GetScreenHeight() * 0.3333333;
			text5PositionX = GetScreenWidth() * 0.05;
			text5PositionY = GetScreenHeight() * 0.95;
			text6PositionX = halfScreenX - MeasureText(text6, sizeText2) / 2;
			text6PositionY = halfScreenY + GetScreenHeight() * 0.1333333;

			colorRect1 = PURPLE;
			colorRect2 = PURPLE;
			colorRect3 = PURPLE;
			colorRect4 = PURPLE;

			rect1.height = (GetScreenWidth() * 80) / scaleAux2;
			rect1.width = (GetScreenWidth() * 255) / scaleAux2;
			rect1.x= halfScreenX - rect1.width / 2;
			rect1.y= halfScreenY + GetScreenHeight() * 0.11;

			rect2.height = (GetScreenWidth() * 80) / scaleAux2;
			rect2.width = (GetScreenWidth() * 255) / scaleAux2;
			rect2.x = halfScreenX - rect1.width / 2;
			rect2.y = halfScreenY + GetScreenHeight() * 0.21;

			rect3.height = (GetScreenWidth() * 80) / scaleAux2;
			rect3.width = (GetScreenWidth() * 255) / scaleAux2;
			rect3.x = halfScreenX - rect1.width / 2;
			rect3.y = halfScreenY + GetScreenHeight() * 0.31;

			rect4.height = (GetScreenWidth() * 80) / scaleAux2;
			rect4.width = (GetScreenWidth() * 255) / scaleAux2;
			rect4.x = halfScreenX - rect1.width / 2;
			rect4.y = halfScreenY + GetScreenHeight() * 0.01;

			scaleBackground= (GetScreenWidth() * 1.0f) / scaleAux3;

		}

		void UpdateMenu()
		{
			mousePoint = GetMousePosition();
			
			if (CheckCollisionPointRec(mousePoint, rect1))
			{
				colorRect1.a = 120;

				if (IsMouseButtonPressed(MOUSE_LEFT_BUTTON))
				{
					mute = !mute;
					cambiarTexto = !cambiarTexto;
				}
			}
			else colorRect1.a = 255;

			if (CheckCollisionPointRec(mousePoint, rect3))
			{
				colorRect3.a = 120;

				if (IsMouseButtonPressed(MOUSE_LEFT_BUTTON)) exit = true;
			}
			else colorRect3.a = 255;

			if (CheckCollisionPointRec(mousePoint, rect2))
			{
				colorRect2.a = 120;

				if (IsMouseButtonPressed(MOUSE_LEFT_BUTTON))
				{
					UnloadMenu();
					currentScreen = Credits;
				}
			}
			else colorRect2.a = 255;
			if (CheckCollisionPointRec(mousePoint, rect4))
			{
				colorRect4.a = 120;

				if (IsMouseButtonPressed(MOUSE_LEFT_BUTTON))
				{
					UnloadMenu();
					InitValues();
					currentScreen = Gameplay;
				}
			}
			else colorRect4.a = 255;

			
		}

		void DrawMenu()
		{
			ClearBackground(WHITE);
			DrawTextureEx(menuImageTexture, { 0,0 }, 0, scaleBackground, WHITE);
			DrawRectangleRec(rect1, colorRect1);
			DrawRectangleRec(rect2, colorRect2);
			DrawRectangleRec(rect3, colorRect3);
			DrawRectangleRec(rect4, colorRect4);
			DrawText(text1, text1PositionX, text1PositionY, sizeText2, BLACK);
			if (!cambiarTexto)
			{
				DrawText(text2, text2PositionX, text2PositionY, sizeText2, BLACK);
			}
			else
				DrawText(text6, text6PositionX, text6PositionY, sizeText2, BLACK);
			
			DrawText(text3, text3PositionX, text3PositionY, sizeText2, BLACK);
			DrawText(text4, text4PositionX, text4PositionY, sizeText2, BLACK);
			DrawText(text5, text5PositionX, text5PositionY, sizeText2, WHITE);
		}

		void UnloadMenu()
		{
			UnloadTexture(menuImageTexture);
			UnloadImage(menuImage);
		}
	}
}

src/Objetos/meteors.cpp
#include "meteors.h"

#include <cmath>

#include "spaceship.h"
#include "game.h"
#include "app.h"

using namespace app;
using namespace game;
using namespace spaceship;

namespace app
{
	namespace meteors
	{
		struct Canon
		{
			Rectangle rec;
			Vector2 position;
			float angle;
			Shoot shoot;
			float timer;
			float shootSpeed;
			float shootDirection;
		};

		static const int maxBigMeteors = 6;
		static Meteor bigMeteor[maxBigMeteors];

		static const int maxFloorMeteors = 5;
		static Meteor floorMeteors[maxFloorMeteors];
		static Canon canons[maxFloorMeteors];
		static Texture2D spriteFM; //floor meteors
		static Texture2D spriteFMS; //floor meteors shoot
		static Vector2 vDirection; //determina la direccion del disparo de las maquinas
		static int destroyedFloorMeteorsCount;

		static int destroyedMeteorsCount;
		static float meteorSpeed;
		static const float speedBoost = 100.0f; //variable para que la velocidad se ajuste al GetFrameTime
		static int initialCorrection = 150;//para que los meteoros no tomen la posicion de la nave en el inicio
		static float posx, posy;
		static float velx, vely;
		static bool correctRange;

		//Imagenes y texturas
		static Image meteorImage;
		static Texture2D meteorTexture;

		//Numeros magicos para escalar las textura
		static float bigMeteorScale;
		static Vector2 bigMeteorScalePos;

		//Sonidos 
		static Sound explosionSound;
		bool pauseSoundExplosion=false;

		static bool init;
		static int scaleAux = 1600;
		static void initFloorMeteors();
		static void updateFloorMeteors();
		static void drawFloorMeteors();
		static const float floorMeteorsDistance = 400 * 6 / maxFloorMeteors;
		const int posMachinesY = 695;

		static void initFloorMeteors()
		{
			destroyedFloorMeteorsCount = 0;
			floorMeteors[0].position = { (float)GetScreenWidth() / 6 * 4, posMachinesY };
			for (int i = 0; i < maxFloorMeteors; i++)
			{
				floorMeteors[i].active = true;
				floorMeteors[i].radius = 50.0f;
				floorMeteors[i].color = RED;
				floorMeteors[i].speed = { 100.0f * 6 / maxFloorMeteors,0.0f };
				if (i != 0)
				{
					floorMeteors[i].position = { floorMeteors[i - 1].position.x + floorMeteorsDistance,posMachinesY };
				}
				floorMeteors[i].tag = i;
				//-----------------------------------

				canons[i].rec.x = floorMeteors[i].position.x;
				canons[i].rec.y = floorMeteors[i].position.y;
				canons[i].rec.width = floorMeteors[i].radius*3/2;
				canons[i].rec.height = 24;
				canons[i].angle = -90;

				spriteFM = LoadTexture("res/assets/machines.png");
				spriteFMS = LoadTexture("res/assets/shoot2.png");

				canons[i].timer = 0.0f;
				canons[i].shoot.active = false;
				canons[i].shoot.position.x = canons[i].rec.x;
				canons[i].shoot.position.y = canons[i].rec.y;
				canons[i].shoot.rotation = canons[i].angle;
				canons[i].shootSpeed = 400.0f;
				canons[i].shoot.radius = 3;
			}
		}

		static int lastMeteorX;

		static void checkFloorMeteorsCol()
		{
			for (int i = 0; i < shipMaxBombs; i++)
			{
				for (int j = 0; j < maxFloorMeteors; j++)
				{
					if (CheckCollisionCircles(bombs[i].position, bombs[i].radius, floorMeteors[j].position, floorMeteors[j].radius) && floorMeteors[j].active)
					{
						floorMeteors[j].active = false;
						bombs[i].active = false;
						destroyedFloorMeteorsCount++;
					}
				}
			}
		}

		static void checkLastMeteor()
		{
			lastMeteorX = 0;
			for (int k = 0; k < maxFloorMeteors; k++)
			{
				if (floorMeteors[k].position.x > lastMeteorX)
				{
					lastMeteorX = floorMeteors[k].position.x;
				}
			}
		}

		static void updateCanonShoots()
		{
			int angleFixed = 79;
			for (int i = 0; i < maxFloorMeteors; i++)
			{
				canons[i].shoot.rotation = canons[i].angle + angleFixed;

				if (!canons[i].shoot.active)
				{
					canons[i].timer += GetFrameTime();
				}
				if (canons[i].timer > 3)
				{
					canons[i].shoot.active = true;
					canons[i].timer = 0;
				}
				if (!canons[i].shoot.active)
				{
					canons[i].shootDirection = canons[i].shoot.rotation;
					canons[i].shoot.position.x = canons[i].rec.x;
					canons[i].shoot.position.y = canons[i].rec.y;
				}
				else
				{
					canons[i].shoot.position.x += sin(canons[i].shootDirection*DEG2RAD)*canons[i].shootSpeed*GetFrameTime();
					canons[i].shoot.position.y -= cos(canons[i].shootDirection*DEG2RAD)*canons[i].shootSpeed*GetFrameTime();
				}
				if (canons[i].shoot.position.x<0 || canons[i].shoot.position.x>GetScreenWidth() ||
					canons[i].shoot.position.y<0 || canons[i].shoot.position.y>GetScreenHeight())
				{
					canons[i].shoot.active = false;
				}

				if (CheckCollisionCircles({ ship.collider.x, ship.collider.y }, ship.collider.z, canons[i].shoot.position, canons[i].shoot.radius))
				{
					gameOver = true;
					victory = false;
					currentScreen = GameOver;
				}
			}
		}
		static void updateFloorMeteors()
		{
			for (int i = 0; i < maxFloorMeteors; i++)
			{
				if (floorMeteors[i].active)
				{
					floorMeteors[i].position.x -= floorMeteors[i].speed.x*GetFrameTime();
					if (floorMeteors[i].position.x < -floorMeteors[i].radius * 2)
					{
						checkLastMeteor();
						floorMeteors[i].position.x = lastMeteorX + floorMeteorsDistance;
					}
					//--------------------------------------
					canons[i].rec.x = floorMeteors[i].position.x;
					canons[i].rec.y = floorMeteors[i].position.y;

					vDirection.x = ship.position.x-canons[i].rec.x;

					vDirection.y = ship.position.y-canons[i].rec.y;

					canons[i].angle = atan2(vDirection.y, vDirection.x)*RAD2DEG;
					//----------------------------------------

					
				}
			}
			updateCanonShoots();
			checkFloorMeteorsCol();
		}

		static void drawFloorMeteors()
		{
			for (int i = 0; i < maxFloorMeteors; i++)
			{
				if (floorMeteors[i].active)
				{
					DrawTexture(spriteFMS, canons[i].shoot.position.x - spriteFMS.width / 2,
						canons[i].shoot.position.y - spriteFMS.height / 2, WHITE);
					DrawTexturePro(spriteFM, { 0.0f,0.0f,(float)spriteFM.width / 2,(float)spriteFM.height },
						{ floorMeteors[i].position.x,floorMeteors[i].position.y,(float)spriteFM.width/2,(float)spriteFM.height },
						{ (float)spriteFM.width / 4,(float)spriteFM.height / 2 }, 0.0f, WHITE);
					DrawTexturePro(spriteFM, { (float)spriteFM.width / 2,0.0f,(float)spriteFM.width / 2,(float)spriteFM.height },
						{floorMeteors[i].position.x,floorMeteors[i].position.y,(float)spriteFM.width/4,(float)spriteFM.height /2},
						{ 0.0f, (float)spriteFM.height / 4 }, canons[i].angle, WHITE);
				}
			}
		}

		void InitMeteors()
		{
			explosionSound = LoadSound("res/assets/explosion.wav");
			meteorImage = LoadImage("res/assets/asteroide.png");
	
			meteorTexture = LoadTextureFromImage(meteorImage);

			correctRange = false;
			init = false;

			destroyedMeteorsCount = 0;
			meteorSpeed = 2;

			for (int i = 0; i < maxBigMeteors; i++)
			{
				posx = GetRandomValue(GetScreenWidth()-bigMeteor[i].radius,GetScreenWidth()-500);

				posy = GetRandomValue(0+ bigMeteor[i].radius*2, GetScreenHeight()- bigMeteor[i].radius*2);

				bigMeteor[i].position = { posx, posy };

				velx = GetRandomValue(-meteorSpeed, -meteorSpeed-1);
				vely = GetRandomValue(-meteorSpeed, meteorSpeed);

				while (!correctRange)
				{
					if (velx == 0 && vely == 0)
					{
						velx = -meteorSpeed;
						vely = GetRandomValue(-meteorSpeed, meteorSpeed);
					}
					else correctRange = true;
				}

				bigMeteor[i].speed = { velx, vely };
				bigMeteor[i].radius = (50 * GetScreenWidth()) / scaleAux;
				bigMeteor[i].active = true;
				bigMeteor[i].color = BLUE;
			}

			bigMeteorScale = (GetScreenWidth()* 0.25f) / scaleAux;
			bigMeteorScalePos = { (bigMeteorScale*meteorImage.width) / 2 ,(bigMeteorScale*meteorImage.height) / 2 };
			initFloorMeteors();
		}

		void UpdateMeteors()
		{
			if (!gameOver)
			{
				for (int a = 0; a < maxBigMeteors; a++)
				{
					if (CheckCollisionCircles({ ship.collider.x, ship.collider.y }, ship.collider.z, bigMeteor[a].position, bigMeteor[a].radius) && bigMeteor[a].active)
					{
						gameOver = true;
						victory = false;
						currentScreen = GameOver;
					}
				}

				// Meteors logic: big meteors
				for (int i = 0; i < maxBigMeteors; i++)
				{
					if (bigMeteor[i].active)
					{
						// Movement
						bigMeteor[i].position.x += bigMeteor[i].speed.x*GetFrameTime() * speedBoost;
						bigMeteor[i].position.y += bigMeteor[i].speed.y*GetFrameTime() * speedBoost;

						// Collision logic: meteor vs wall
						if (bigMeteor[i].position.x > GetScreenWidth() + bigMeteor[i].radius)
						{
							bigMeteor[i].position.x = -(bigMeteor[i].radius);
						}
						else if (bigMeteor[i].position.x < 0 - bigMeteor[i].radius)
						{
							bigMeteor[i].position.x = GetScreenWidth() + bigMeteor[i].radius;
							bigMeteor[i].position.y= GetRandomValue(bigMeteor[i].radius, GetScreenHeight());
						}
						if (bigMeteor[i].position.y > GetScreenHeight() - bigMeteor[i].radius)
						{
							bigMeteor[i].speed.y *= -1;
						}
						else if (bigMeteor[i].position.y < 0 + bigMeteor[i].radius)
						{
							bigMeteor[i].speed.y *= -1;
						}
					}
				}

				// Collision logic: player-shoots vs meteors
				for (int i = 0; i < shipMaxShoots; i++)
				{
					if ((shoot[i].active))
					{
						for (int a = 0; a < maxBigMeteors; a++)
						{
							if (bigMeteor[a].active && CheckCollisionCircles(shoot[i].position, shoot[i].radius, bigMeteor[a].position, bigMeteor[a].radius))
							{
								if (!pauseSoundExplosion)
								{
									PlaySound(explosionSound);
								}
								else
								{
									PauseSound(explosionSound);
								}
								shoot[i].active = false;
								bigMeteor[a].active = false;
								destroyedMeteorsCount++;
								a = maxBigMeteors;
							}
						}

						
					}
				}
			}

			updateFloorMeteors();

			if (destroyedMeteorsCount == maxBigMeteors && destroyedFloorMeteorsCount==maxFloorMeteors)
			{
				victory = true;
				gameOver = true;
				currentScreen = GameOver;
			}

		}

		void DrawMeteors()
		{
			for (int i = 0; i < maxBigMeteors; i++)
			{
				if (bigMeteor[i].active)
				{
					DrawCircleV(bigMeteor[i].position, bigMeteor[i].radius, DARKGRAY);
					DrawTextureEx(meteorTexture, { bigMeteor[i].position.x - bigMeteorScalePos.x,bigMeteor[i].position.y - bigMeteorScalePos.y }, 0, bigMeteorScale, WHITE);
				}
				else DrawCircleV(bigMeteor[i].position, bigMeteor[i].radius, BLANK);
			}
			drawFloorMeteors();
		}

		void UnloadMeteors()
		{
			UnloadSound(explosionSound);
			UnloadTexture(meteorTexture);
			UnloadImage(meteorImage);
			UnloadTexture(spriteFM);
			UnloadTexture(spriteFMS);
		}

		void ResetMeteors()
		{
			correctRange = false;
			init = false;

			destroyedMeteorsCount = 0;
			meteorSpeed = 2;

			for (int i = 0; i < maxBigMeteors; i++)
			{
				posx = GetRandomValue(GetScreenWidth() - bigMeteor[i].radius, GetScreenWidth() - 500);

				posy = GetRandomValue(0 + bigMeteor[i].radius * 2, GetScreenHeight() - bigMeteor[i].radius * 2);

				bigMeteor[i].position = { posx, posy };

				velx = GetRandomValue(-meteorSpeed, -meteorSpeed - 1);
				vely = GetRandomValue(-meteorSpeed, meteorSpeed);

				while (!correctRange)
				{
					if (velx == 0 && vely == 0)
					{
						velx = -meteorSpeed;
						vely = GetRandomValue(-meteorSpeed, meteorSpeed);
					}
					else correctRange = true;
				}

				bigMeteor[i].speed = { velx, vely };
				bigMeteor[i].radius = (50 * GetScreenWidth()) / scaleAux;
				bigMeteor[i].active = true;
				bigMeteor[i].color = BLUE;
			}

			bigMeteorScale = (GetScreenWidth()* 0.25f) / scaleAux;
			bigMeteorScalePos = { (bigMeteorScale*meteorImage.width) / 2 ,(bigMeteorScale*meteorImage.height) / 2 };
			initFloorMeteors();
		}
	}
}

src/Objetos/spaceship.h
#ifndef SPACESHIP_H
#define SPACESHIP_H

#include <raylib.h>

namespace app
{
	namespace spaceship
	{
		void InitSpaceship();
		void UpdateSpaceship();
		void DrawSpaceship();
		void UnloadSpaceship();
		void ResetSpaceship();

		struct Spaceship
		{
			Vector2 position;
			Vector2 speed;
			Vector2 acceleration;
			float rotation;
			Vector3 collider;
			Color color;
		};

		struct Shoot
		{
			Vector2 position;
			Vector2 speed;
			float radius;
			float rotation;
			bool active;
			Color color;
		};
		
		struct Bomb
		{
			Vector2 position;
			Vector2 speed;
			float radius;
			float rotation;
			bool active;
			Color color;
			int tag;
		};

		extern Bomb bombs[];
		extern Spaceship ship;
		extern const int shipMaxShoots;
		extern const int shipMaxBombs;
		extern Shoot shoot[];
		extern bool pauseSoundShoot;
	}
}
#endif // !SPACESHIP_H


src/Pantallas/game.cpp
#include "game.h"

#include <raylib.h>

#include "meteors.h"
#include "spaceship.h"
#include "app.h"

using namespace app;
using namespace	spaceship;
using namespace	meteors;

namespace app
{
	namespace game
	{
		extern bool victory = false;
		extern bool gameOver = false;
		static char text[] = "W A S D : movimiento, CLICK IZQ: Disparar, CLICK DER: Lanzar Bomba";
		static int sizeText;
		static float textPositionX;
		static float textPositionY;

		//Imagenes y texturas

		static Image backImage;
		static Texture2D backTexture;
		static Image backImage2;
		static Texture2D backTexture2;

		static Rectangle frameRec;
		static float currentFrame;

		static float framesCounter;
		static float framesSpeed;

		static Rectangle frameRec2;
		static float currentFrame2;

		static float framesCounter2;
		static float framesSpeed2;

		//Boton pausa

		static Rectangle botonPausa1;
		static Rectangle botonPausa2;
		static Color colorRect;
		static Vector2 mousePoint;

		//Puntaje 

		int score;
		

		static bool init;

		void InitValues()
		{
			sizeText = (GetScreenWidth() * 22) / 1600;
			textPositionX = GetScreenWidth()*0.01f;
			textPositionY = GetScreenHeight() * 0.97f;

			backImage = LoadImage("res/assets/parallax.png");
			backImage2 = LoadImage("res/assets/parallax2.png");

			//init boton pausa
			botonPausa1.x = GetScreenWidth()*0.96f;
			botonPausa1.y = GetScreenHeight()*0.02f;
			botonPausa1.height = (GetScreenWidth() * 40) / 1600;
			botonPausa1.width = (GetScreenWidth() * 15) / 1600;
			botonPausa2.x = GetScreenWidth()*0.975f;
			botonPausa2.y = GetScreenHeight()*0.02f;
			botonPausa2.height = (GetScreenWidth() * 40) / 1600;
			botonPausa2.width = (GetScreenWidth() * 15) / 1600;
			colorRect = GRAY;

			backTexture = LoadTextureFromImage(backImage);
			backTexture2 = LoadTextureFromImage(backImage2);

			frameRec = { 0.0f, 0.0f, (float)backTexture.width, (float)backTexture.height };
			currentFrame = 0;

			framesCounter = 0;
			framesSpeed = 0.001f;

			frameRec2 = { 0.0f, 0.0f, (float)backTexture2.width, (float)backTexture2.height };
			currentFrame2 = 0;

			framesCounter2 = 0;
			framesSpeed2 = 0.001f;

			score = 0;
			
			InitSpaceship();
			InitMeteors();
		}
		

		static void Input()
		{
			mousePoint = GetMousePosition();
			if (CheckCollisionPointRec(mousePoint, botonPausa1))
			{
				colorRect.a = 120;

				if (IsMouseButtonPressed(MOUSE_LEFT_BUTTON))
				{
					currentScreen = Pause;
				}
			}
			else colorRect.a = 255;

			if (CheckCollisionPointRec(mousePoint, botonPausa2))
			{
				if (IsMouseButtonPressed(MOUSE_LEFT_BUTTON))
				{
					currentScreen = Pause;
				}
			}

		}

		

		static void Update()
		{
			framesCounter+=GetFrameTime();

			if (framesCounter >= framesSpeed)
			{
				framesCounter = 0;
				currentFrame+=0.1;

				if (currentFrame > 1594) currentFrame = 0;

				frameRec.x = (float)currentFrame;
			}
			framesCounter2+=GetFrameTime();

			if (framesCounter2 >= framesSpeed2)
			{
				framesCounter2 = 0;
				currentFrame2 += 1;

				if (currentFrame2 > 1280) currentFrame2 = 0;

				frameRec2.x = (float)currentFrame2;
			}

			UpdateSpaceship();
			UpdateMeteors();
		}

		void UpdateFrame()
		{
			Input();
			Update();
		}

		void DrawGameplay()
		{
			ClearBackground(BLANK);
			DrawTextureRec(backTexture, frameRec, { 0,0 }, WHITE);
			DrawSpaceship();
			DrawMeteors();
			DrawTextureRec(backTexture2, frameRec2, { 0,0 }, WHITE);


			DrawRectangleRec(botonPausa1, colorRect);
			DrawRectangleRec(botonPausa2, colorRect);
			DrawText(text, textPositionX, textPositionY, sizeText, PURPLE);
		}

		void ResetValues()
		{
			sizeText = (GetScreenWidth() * 20) / 1600;
			textPositionX = GetScreenWidth()*0.01f;
			textPositionY = GetScreenHeight() * 0.97f;

			//init boton pausa
			botonPausa1.x = GetScreenWidth()*0.96f;
			botonPausa1.y = GetScreenHeight()*0.02f;
			botonPausa1.height = (GetScreenWidth() * 40) / 1600;
			botonPausa1.width = (GetScreenWidth() * 15) / 1600;
			botonPausa2.x = GetScreenWidth()*0.975f;
			botonPausa2.y = GetScreenHeight()*0.02f;
			botonPausa2.height = (GetScreenWidth() * 40) / 1600;
			botonPausa2.width = (GetScreenWidth() * 15) / 1600;
			colorRect = GRAY;

			frameRec = { 0.0f, 0.0f, (float)backTexture.width, (float)backTexture.height };
			currentFrame = 0;

			framesCounter = 0;
			framesSpeed = 0.001f;

			frameRec2 = { 0.0f, 0.0f, (float)backTexture2.width, (float)backTexture2.height };
			currentFrame2 = 0;

			framesCounter2 = 0;
			framesSpeed2 = 0.001f;

			score = 0;

			ResetSpaceship();
			ResetMeteors();
			gameOver = false;
		}

		void UnloadGameplay()
		{
			UnloadMeteors();
			UnloadSpaceship();
			UnloadTexture(backTexture);
			UnloadTexture(backTexture2);
			UnloadImage(backImage);
			UnloadImage(backImage2);
		}
	}
}